### ISSUE TITLE
ValueError with pad_list[i] = -1 in an unfortunate coincidence of how traces are sampled

### ISSUE THREAD

**Issue by flixha** [eqcorrscan/EQcorrscan#425]
**Describe the bug**
[This line in `correlate`](https://github.com/eqcorrscan/EQcorrscan/blob/aadaf9b184f5ef52928f8a13e2df7eb553a63124/eqcorrscan/utils/correlate.py#L490) can cause a `ValueError` due to a negative dimenstion when an element in `pad_list` becomes `-1`.

**To Reproduce**
It seems easiest to describe with a theoretically (rather than uploading files etc.), so consider the following values:
```python
stream_channel.stats.sampling_rate = 20
stream_channel.stats.starttime     = UTCDateTime(2013, 1, 2, 0, 0, 0, 19539)
stream_start                       = UTCDateTime(2013, 1, 1, 23, 59, 59, 991000)

template[i].stats.sampling_rate    = 20
template[i].stats.starttime        = UTCDateTime(2018, 4, 20, 18, 14, 42, 500000)
t_starts[j]                        = UTCDateTime(2018, 4, 20, 18, 14, 42, 500000)

```

With the [code in `_get_array_dicts` (lines 932-942)](https://github.com/eqcorrscan/EQcorrscan/blob/aadaf9b184f5ef52928f8a13e2df7eb553a63124/eqcorrscan/utils/correlate.py#L932-L942) and the values above, we get:

```python
stream_offset = int(round(stream_channel.stats.sampling_rate * 
                    (stream_channel.stats.starttime - stream_start)))
stream_offset = 1
pad_list[i]   = int(round(template[i].stats.sampling_rate * 
                   (template[i].stats.starttime - t_starts[j]))) - stream_offset
pad_list[i]   = -1

```

**Expected behavior**
`pad_list[i]` should never become smaller than 0.

**Desktop (please complete the following information):**
 - Operating System: RHEL 8
 - Python version: 3.7
 - EQcorrscan version: Develop-branch

**Additional context**
This bug appears to be caused by an unfortunate incidence of how the data are sampled. 
Note that the starttimes for `stream` and `stream_channel` are each less than half a sample from midnight. But their difference is just above half a sample (0.028 s at 20 Hz, so 0.57 samples).
It does not matter how one has trimmed the day's stream (~~`nearest_sample=True` or `False`,~~ with `pad=True` or `False`) or how one has downsampled the data.
The only workaround would be to interpolate the day's stream, but I believe that is not entirely necessary considering its computational cost.

I'm writing down a possible solution in the discussion, but I would very much appreciate if someone could have a critical look, think about potential consequences of the suggested solutions below, and/or suggest an alternative solution.

**Comment by flixha** [comment]
If I understand correctly, slight sampling offsets of less than half a sample on each trace aren't really problematic when correlating multiple channels. So the solution should be relatively simple, by just avoiding `pad = -1`.

Possible solutions: 

1. Always round down the stream_offset:
```python
stream_offset = int(math.floor((stream_channel.stats.sampling_rate * 
                    (stream_channel.stats.starttime - stream_start)))
```

2. Only take care of `pad=-1`, and manually set these to 0 before adding `pads`to :
```python
pads = [0 if p==-1 else p for p in pads]
```

**Comment by flixha** [comment]
So for this very occurrence where I got the error, manually trimming the data with `nearest_sample=False` did after all stop the error from occurring (but only in this instance, and I'm not sure this can be generalized).

However, the default in eqcorrscan's [`pre_processing`](https://github.com/eqcorrscan/EQcorrscan/blob/aadaf9b184f5ef52928f8a13e2df7eb553a63124/eqcorrscan/utils/pre_processing.py#L614-L615) is to use `nearest_sample=True`, in which case the error definitely can occur.

**Comment by calum-chamberlain** [comment]
Thanks for reporting this @flixha and for the detailed information and suggestions - this is an excellent bug report!
I suspect that option one (using floor) is probably the way to go.  The pads don't mater for the actual correlations, but are used in the stacking.

I will think a bit more about this and the implications for stacking the cross-correlations. I think we should make a test for `utils.correlate` that has a stream with these "unfortunate" time-stamps and make sure that we get what we expect. Synthetic data should be good enough for that.

**Comment by calum-chamberlain** [comment]
@flixha is this fixed in one of your PRs or is this still to go?

**Comment by flixha** [comment]
Hi @calum-chamberlain, thanks for the push to get all the issues resolved and getting ready for 0.4.3! Sorry that I wasn't quick to reply to some open issues since Tuesday, as I'm only on mobile for a few days.
I see that the fix to the issue here is now already in develop, it's in `correlate._get_array_dicts`. I can write a test for this issue after the weekend.

**Comment by calum-chamberlain** [comment]
All good - a test would be great, but I didn't mean to pester you, sorry! Thanks for all your work fixing things!
